### PR TITLE
[YUNIKORN-352] - fixed max resource check and behavioural change

### PR DESCRIPTION
**Pull request by kingamarton**

### CODE REVIEW

**Review by wilfred-s — CHANGES_REQUESTED** [review]
I think we can simplify the checks in the hierarchy by doing the checks two directions.
* max check on the way down into the recursion (parent > child)
* guaranteed check on the way up out of the recursion (child sum < parent)

File: pkg/common/configs/configvalidator.go [hunk 1]
```diff
@@ -75,65 +75,100 @@ func checkACL(acl string) error {
...
+			return err
 		}
+	}
 
-		if cur.Resources.Guaranteed != nil {
-			if !resources.FitIn(curGuaranteedRes, sum) {
-				return fmt.Errorf("queue %s has guaranteed-resources (%v) smaller than sum of children guaranteed resources (%v)", cur.Name, curGuaranteedRes, sum)
-			}
-		} else {
-			cur.Resources.Guaranteed = sum.ToConf()
```
[the remark this answers is not in the document]
Comment (kingamarton) [reply pkg/common/configs/configvalidator.go:118]: reverted that func

File: pkg/common/configs/configvalidator.go [hunk 2]
```diff
@@ -75,65 +75,100 @@ func checkACL(acl string) error {
 // Check the queue resource configuration settings.
 // - child or children cannot have higher maximum or guaranteed limits than parents
 // - children (added together) cannot have a higher guaranteed setting than a parent
-func checkResourceConfigurationsForQueue(cur *QueueConfig, parent *QueueConfig) error {
+func checkResourceConfigurationsForQueue(cur QueueConfig, parentMax *resources.Resource, parentGuaranteed *resources.Resource) error {
```
[the remark this answers is not in the document]
Comment (kingamarton) [reply pkg/common/configs/configvalidator.go]: Removed repeated  parameter type

File: pkg/common/resources/resources.go [hunk 3]
```diff
@@ -771,6 +803,13 @@ func IsZero(zero *Resource) bool {
 	return true
 }
 
+func IsEmpty(r *Resource) bool {
+	if r == nil {
+		return true
+	}
+	return len(r.Resources) == 0
+}
+
```
[the remark this answers is not in the document]
Comment (kingamarton) [reply pkg/common/resources/resources.go]: Removed this method and extended isZero check with checking if the resources are empty and if yes handle it as zero resource

File: pkg/common/resources/resources.go [hunk 4]
```diff
@@ -742,6 +743,37 @@ func ComponentWiseMin(left, right *Resource) *Resource {
...
 }
 
+// Returns a new Resource with the smallest value for each quantity in the Resources
+// If either Resource passed in is nil the other Resource is returned
+// If a Resource type is missing from one of the Resource, it is considered empty and the quantity from the other Resource is returned
+func ComponentWiseMinPermissive(left, right *Resource) *Resource {
+	out := NewResource()
+	if right == nil && left == nil {
+		return out
+	}
```
[the remark this answers is not in the document]
Comment (kingamarton) [reply pkg/common/resources/resources.go:761]: You are right! Changed to nil

File: pkg/common/resources/resources.go [hunk 5]
```diff
@@ -742,6 +743,37 @@ func ComponentWiseMin(left, right *Resource) *Resource {
...
+	if left == nil {
+		return right
+	}
+	if right == nil {
+		return left
+	}
```
[the remark this answers is not in the document]
Comment (kingamarton) [reply pkg/common/resources/resources.go:767]: Done

**Comment by wilfred-s** [comment]
The patch looks good beside the codecov remarks.
I can see that the error or no input cases are not hit based on the remarks it puts in. Specially in the resources `ComponentWiseMinPermissive()` we can get a couple easily fixed by nil testing. However this is a central part of the code and we should really strive for a higher test coverage.

**Review by wilfred-s — CHANGES_REQUESTED** [review]
test coverage has to be better for error cases, we drop a lot of easily coverable lines

**Review by wilfred-s — APPROVED** [review]
LGTM, additional unit test look good

Change breaks existing unit test just committed from YUNIKORN-466, fix on commit

**Pull request closed without merging**